cornac/models/tifuknn/recom_tifuknn.py

```diff
@@ -0,0 +1,171 @@
+            for iid in iids:
+                his_vec[iid] = decayed_val
+            his_list.append(his_vec)
+        grouped_list, real_n_groups = self._group_history_list(his_list, self.n_groups)
+        his_vec = np.zeros(self.total_items, dtype="float32")
+        if real_n_groups == 0:
+            return his_vec
+
+        for idx in range(real_n_groups):
+            decayed_val = np.power(self.group_decay_rate, self.n_groups - idx - 1)
+            his_vec += grouped_list[idx] * decayed_val
+
+        return his_vec / real_n_groups
+
+    def _group_history_list(self, his_list, n_groups):
+        grouped_vec_list = []
+        if len(his_list) < n_groups:
+            for j in range(len(his_list)):
+                grouped_vec_list.append(his_list[j])
+            return grouped_vec_list, len(his_list)
+        else:
+            est_num_vec_each_block = len(his_list) / n_groups
+            base_num_vec_each_block = int(np.floor(len(his_list) / n_groups))
+            residual = est_num_vec_each_block - base_num_vec_each_block
+
+            num_vec_has_extra_vec = int(np.round(residual * n_groups))
+
+            if residual == 0:
+                for i in range(n_groups):
+                    sum = np.zeros(len(his_list[0]))
+                    for j in range(base_num_vec_each_block):
+                        sum += his_list[i * base_num_vec_each_block + j]
+                    grouped_vec_list.append(sum / base_num_vec_each_block)
+            else:
+                for i in range(n_groups - num_vec_has_extra_vec):
+                    sum = np.zeros(len(his_list[0]))
+                    for j in range(base_num_vec_each_block):
+                        sum += his_list[i * base_num_vec_each_block + j]
+                        last_idx = i * base_num_vec_each_block + j
+                    grouped_vec_list.append(sum / base_num_vec_each_block)
+
+                est_num = int(np.ceil(est_num_vec_each_block))
+                start_group_idx = n_groups - num_vec_has_extra_vec
+                if len(his_list) - start_group_idx * base_num_vec_each_block >= est_num_vec_each_block:
+                    for i in range(start_group_idx, n_groups):
+                        sum = np.zeros(len(his_list[0]))
+                        for j in range(est_num):
+                            iidxx = last_idx + 1 + (i - start_group_idx) * est_num + j
+                            sum += his_list[iidxx]
+                        grouped_vec_list.append(sum / est_num)
+
+            return grouped_vec_list, n_groups
+
+    def score(self, user_idx, history_baskets, **kwargs):
+        if len(history_baskets) == 0:
+            return np.zeros(self.total_items, dtype="float32")
+        user_vector = self._compute_user_vector(history_baskets)
+        _, indices = self.tree.query([user_vector], k=self.n_neighbors)
+        return self.alpha * user_vector + (1 - self.alpha) * np.sum([self.user_vectors[idx] for idx in indices.squeeze()], axis=0)
```

Comment:
Based on Section 3.1 in the paper, $\textbf{u}_n$ should be the average vector of neighbors, not sum. Please double check on that. Also, we simply pass the indices directly instead of doing for loop:
```python
np.mean(self.user_vectors[indices.squeeze()])
```

Reply:
@lthoang

Reply:
I've  double checked. It should be the average vector of neighbors. I will make update in the next commit.